### PR TITLE
feat(e2e): add step driver registry and coverage reporting

### PR DESCRIPTION
**Pull request by leslie-heinzen**
## Summary

This PR adds a step-driver registry to the E2E runner and reports coverage for every rendered step kind.

It builds on #1640, which added the tracked step-root contract for current DOM discovery.

A guide with only unsupported roots now returns a schema-valid skipped report. The report keeps its full coverage inventory and explicit reason.

## What to look at

- [Driver registry](https://github.com/grafana/grafana-pathfinder-app/blob/f72c8c3256f5ecc8b6e4e1d4b1d9057bb3686aaa/tests/e2e-runner/utils/guide-runner/drivers/registry.ts): Confirm that all nine step kinds have entries. `plain`, `multistep`, and `guided` support execution. The other six kinds are reporting-only.
- [Discovery and execution](https://github.com/grafana/grafana-pathfinder-app/blob/f72c8c3256f5ecc8b6e4e1d4b1d9057bb3686aaa/tests/e2e-runner/utils/guide-runner/discovery.ts): Confirm that discovery and execution use the registry without step-kind branches outside it.
- [E2E report schema](https://github.com/grafana/grafana-pathfinder-app/blob/f72c8c3256f5ecc8b6e4e1d4b1d9057bb3686aaa/src/cli/e2e/schemas/e2e-report.schema.ts): Confirm that optional `coverage` and `steps[].stepKind` fields preserve the `1.1.0` report contract.
- [Runner integration](https://github.com/grafana/grafana-pathfinder-app/blob/f72c8c3256f5ecc8b6e4e1d4b1d9057bb3686aaa/tests/e2e-runner/utils/guide-runner/run-guide.ts): Confirm that mixed guides report unsupported coverage. Unsupported-only guides return skipped reports with named kinds and complete coverage.
- [CLI result handling](https://github.com/grafana/grafana-pathfinder-app/blob/f72c8c3256f5ecc8b6e4e1d4b1d9057bb3686aaa/src/cli/e2e/e2e-results.ts): Confirm that unsupported-only reports use the skipped status and a successful process exit.

## Why

The runner previously exercised only some rendered step kinds and did not show this coverage gap.

The registry gives later drivers one extension point. Coverage reports show unsupported kinds before those drivers exist.

Unsupported roots do not fail mixed guides. A guide with only unsupported roots returns a skipped report before execution.

This result uses exit code 0, keeps coverage, and blocks guides that declare it as a prerequisite.

Separate kind checks in discovery and execution would split this contract across files and hide missing support.

## How to verify

1. Run a guide with `plain`, `multistep`, and `guided` steps. Confirm that the runner finds each current root and completes supported actions.
2. Run a mixed guide with one reporting-only step. Confirm that the report lists the step under `unsupportedSteps` and keeps the supported-step outcome.
3. Run a guide with only reporting-only steps. Confirm `outcome: "skipped"`, no `errorCode`, and complete `unsupportedSteps` coverage.
4. Run a dependency chain with that guide as a prerequisite. Confirm that the dependent guide gets `SKIPPED_PREREQ` without a failing process exit.
5. Inspect the multi-guide report. Confirm that the summary counts the unsupported-only guide as skipped, not failed.
6. Run a legacy guide without tracked roots. Confirm that the runner uses the legacy selector only when no current roots exist.

## Breaking changes

None. The report fields are optional, and the schema version remains `1.1.0`. Mixed-guide outcomes and exit codes remain unchanged.

Unsupported-only guides now use the existing skipped outcome. This behavior keeps zero-execution coverage without adding a schema enum.

## Out of scope

- Execution drivers for `quiz`, `terminal`, `terminal-connect`, `codeblock`, `challenge`, and `datasource-check`. Later PRs add these drivers.

Co-Authored-By: Warp <agent@warp.dev>

### CODE REVIEW

File: tests/e2e-runner/utils/guide-runner/run-guide.ts [hunk 1]
```diff
@@ -128,12 +131,12 @@ async function executeGuideSteps(
   timestamp: string,
   monitor: BrowserTerminationMonitor
 ): Promise<TestResultsData> {
-  const firstStep = page.locator(STEP_SELECTOR).first();
+  const firstStep = page.locator(STEP_ROOT_SELECTOR).first();
```
[the remark this answers is not in the document]
Comment (leslie-heinzen) [reply tests/e2e-runner/utils/guide-runner/run-guide.ts]: Fixed in 50682926. The runner now filters `STEP_ROOT_SELECTOR` for visible matches before it selects the first root. A later visible root can now satisfy readiness when an earlier root is hidden. The focused runner tests, typecheck, ESLint, Prettier, and the diff check pass.

**Review by moxious — CHANGES_REQUESTED** [review]
## Merge contract

Fix this item and this PR is mergeable.

1. [blocking] **unsupported-only-guide-passes-green — A guide whose tracked roots are all unsupported kinds now reports outcome passed with zero steps executed** (medium · cli-and-e2e-runner)
   Both docs added here state that unsupported coverage does not change guide outcomes, but it does for one shape. The load gate at run-guide.ts:134 moved to STEP_ROOT_SELECTOR, which matches the tracked roots that #1640 added to all nine step components, and the empty-guide guard at run-guide.ts:139 moved from discovery.totalSteps === 0 to discovery.coverage.rendered === 0. So a guide built only from quiz, terminal, terminal-connect, code-block, challenge or datasource-check blocks now clears both, discovery diverts every root to unsupportedSteps leaving steps empty, summarizeResults([]) returns success true, and the outcome ternary yields passed with errorCode undefined and coverage.executed 0. At base the same guide matched no root, threw at the 15s waitFor, and shared-chain turned that into outcome infrastructure_error with errorCode REPORT_MISSING that also blocked dependent milestones. No bundled guide is in this shape today, but every one of those six block types counts as interactive, so an authorable quiz-only or terminal-only guide bypasses the legitimate read-only zero-block early pass and lands on the green path. run-guide.test.ts:308 pins the new behaviour under a title saying the outcome is unchanged.
   Required: Give rendered > 0 && supported === 0 its own distinguishable result rather than passed: either the existing skipped outcome with an explicit reason, or keep a throw whose message names the unsupported kinds. Then make the two doc sentences and the run-guide test title state the chosen semantics precisely.

## Follow-ups

These are tracked separately and do not block merge.

1. [follow_up] **guided-info-edge-case-tests-dropped — The deleted extractGuidedInfo edge-case tests were not replaced for their successors in the driver registry** (low · testing-and-verification)
   Base discovery.test.ts held eight extractGuidedInfo cases. Head re-covers three of them indirectly through discoverStepsFromDOM, but nothing anywhere tests resolveLegacyStepKind's explicit guided targetAction disjunct at registry.ts:123, inspectGuided's fallback-to-1 branches for missing, non-numeric and zero data-test-substep-total at registry.ts:41-44, or inspectMultistep's action-count fallbacks at registry.ts:20-36. registry.test.ts asserts only the kind-to-driver map and the supported partition. Behaviour was verified byte-equivalent to base, so this is a refactor-time coverage regression rather than a behaviour change, and it sits on the legacy compatibility path that cannot be re-verified against the current build, where a wrong action count silently changes the step timeout budget.
   Follow-up: Restore the deleted edge cases against their successors: resolveLegacyStepKind for an explicit guided targetAction, and inspectGuided plus inspectMultistep for missing, non-numeric, zero and non-array action counts.

2. [follow_up] **e2e-testing-file-map-omits-drivers — The E2E_TESTING.md file map still describes guide-runner as it was before the drivers directory existed** (low · cross-cutting-architecture)
   E2E_TESTING.md:26 describes tests/e2e-runner/utils/guide-runner/ as step discovery, execution, browser-termination monitoring, requirement fixing, artifact capture and failure classification, naming no subdirectory. This PR removes 825 lines from execution.ts and moves metadata inspection, per-step timeout, completion, skip and execution into the new drivers/ directory. The PR edits three other hunks of this same file but not the map, and the map already itemises milestone-replacement.ts as a peer entry, so the omission is inconsistent with its own convention. A reader looking for where step behaviour lives is still pointed at execution.ts, which no longer holds it.
   Follow-up: Add a tests/e2e-runner/utils/guide-runner/drivers/ line to the file map naming the registry as the per-step-kind extension point, alongside the existing milestone-replacement.ts entry.



PR Review: https://github.com/grafana/grafana-pathfinder-app/pull/1834
Summary: add step driver registry and coverage reporting
Verdict: Request Changes
Results: 1 blocker, 0 non-blocking findings, 2 follow-ups

**Comment by leslie-heinzen** [comment]
@moxious Addressed the blocking `unsupported-only-guide-passes-green` concern in 29f70655. Unsupported-only guides now fail before execution after DOM discovery. The error names sorted, unique unsupported kinds. Mixed guides retain coverage and their existing outcomes. Both E2E documents and the regression test now state this contract. Validation passed: 40 suites with 561 tests, typecheck, ESLint, Prettier, and `git diff --check`.

File: tests/e2e-runner/utils/guide-runner/run-guide.ts [hunk 2]
```diff
@@ -128,21 +131,26 @@ async function executeGuideSteps(
...
   const guideTimeout = calculateGuideTimeout(discovery.steps);
   options.onTimeoutCalculated?.(guideTimeout);
-  if (discovery.totalSteps === 0) {
+  if (discovery.coverage.rendered === 0) {
     throw new Error(`Guide ${guide.id} contains interactive blocks but rendered no interactive steps`);
   }
+  if (discovery.coverage.supported === 0) {
+    const unsupportedKinds = [...new Set(discovery.coverage.unsupportedSteps.map(({ stepKind }) => stepKind))].sort();
+    throw new Error(`Guide ${guide.id} rendered only unsupported step kinds: ${unsupportedKinds.join(', ')}`);
+  }
```
[the remark this answers is not in the document]
Comment (leslie-heinzen) [reply tests/e2e-runner/utils/guide-runner/run-guide.ts]: Fixed in f72c8c32. Unsupported-only guides now return a schema-valid skipped report before execution. The report keeps coverage.unsupportedSteps, omits errorCode, and uses exit code 0. Standalone and shared paths preserve the report. Dependents still receive SKIPPED_PREREQ, and aggregate summaries count the guide as skipped. The focused run passed 40 suites and 569 tests. Type checking, ESLint, Prettier, and git diff --check also passed.

**Comment by leslie-heinzen** [comment]
Created issues for the two non-blocking follow-ups from @moxious’s review:

- #1838 — Restore edge-case tests for guided and multistep E2E drivers
- #1839 — Document the E2E driver registry in the testing source map

**Review by moxious — APPROVED** [review]
Verified that everything previously mentioned was fixed or dealt with.  I won't bore you with the various nits & non-useful complaints of the second round review.  :)
